src/rbac/DefaultRoleManager.h

```diff
@@ -0,0 +1,260 @@
+#include <map>
+
+#include "RoleManager.h"
+#include "exception/IllegalArgumentException.h"
+#include "log/LogUtil.h"
+
+/**
+ * Role represents the data structure for a role in RBAC.
+ */
+class Role {
+    
+    private:
+        vector <Role> roles;
+
+    public:
+        string name;
+
+        Role(string name) {
+            this->name = name;
+        }
+
+        void addRole(Role role) {
+            for (Role r : roles) {
+                if (!r.name.compare(role.name)) {
+                    return;
+                }
+            }
+
+            roles.push_back(role);
+        }
+
+        void deleteRole(Role role) {
+            vector <Role> :: iterator it;
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                if (!(*it).name.compare(role.name)) {
+                    roles.erase(it);
+                }
+            }
+        }
+
+        bool hasRole(string name, int hierarchyLevel) {
+            if (!this->name.compare(name)) {
+                return true;
+            }
+
+            if (hierarchyLevel <= 0) {
+                return false;
+            }
+
+            vector <Role> :: iterator it;
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                if ((*it).hasRole(name, hierarchyLevel - 1)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        bool hasDirectRole(string name) {
+            vector <Role> :: iterator it;
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                if (!(*it).name.compare(name)) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        string toString() {
+            string names = "";
+            for (int i = 0; i < roles.size(); i ++) {
+                Role role = roles[i];
+                if (i == 0) {
+                    names.append(role.name);
+                } else {
+                    names.append(", " + role.name);
+                }
+            }
+            return name + " < " + names;
+        }
+
+        vector<string> getRoles() {
+            vector <string> names;
+            vector <Role> :: iterator it;
+            for (it = roles.begin() ; it != roles.end() ; it++) {
+                names.push_back((*it).name);
+            }
+            return names;
+        }
+};
+
+class DefaultRoleManager : public RoleManager {
+    private:
+        map <string, Role> allRoles;
```

Comment:
use `map` or `unordered_map`?

Reply:
Yeah will use unordered_map here as the order of the keys won't matter here